### ISSUE TITLE
Invalid status for retry tests

### ISSUE THREAD

**Issue by abhibajaj1987** [reportportal/reportportal#2164]
Description:
We are running our testng tests and retrying it via retry listener of testng.. At the same time we are sending the tests results to report portal listener, however observed that for some of the tests RP is showing invalid status. 
As per retry logic, if test is passed in retries, the test should be tagged with PASSED status ( with or without retries) 

Expected Behaviour: 
The test is marked as skipped which is incorrect since the test passed eventually but somehow RP is showing skippd status**

**
<img width="1469" alt="image" src="https://github.com/reportportal/reportportal/assets/148849999/901f51cd-3d82-48f2-8ed3-078e08176ded">
**

**Versions:**
 - MacOS, Chrome
 - Version of RP is 5.8

**Comment by DzmitryHumianiuk** [comment]
Hi @abhibajaj1987, I have to respectfully disagree with you on this matter. 

From a practical standpoint, the outcome often hinges on the specific test runner and the logic behind its implementation.

Typically, most test runners rely more on the actual test outcome rather than a combination of previous statuses. 
Here's what I mean: 
If a particular test has a retry parameter set up, say it's set to retry up to 5 times upon failure, the test engine will keep re-running this test/method until it either receives a `PASSED` status or reaches the maximum retry count of 5. 
Based on this logic, the most recent status (from the last retry) is the most relevant one. So, the final status would be `PASSED` if the test eventually passes within the given retry attempts, or it could be the last status (either `PASSED` or `FAILED`) if it reaches the retry limit.

This means the actual test execution control remains with the test runner (like TestNG in this instance). 
The test runner decides whether to retry a test or if it has already received a `PASSED` result.

In the scenario you're describing, it seems like you want the decision-making to happen on the ReportPortal side. 
When ReportPortal receives an array of statuses and if any of them is `PASSED`, it considers the test as `PASSED`.

Ultimately, the debate boils down to deciding which entity, the test runner or ReportPortal, should have the final say. 
But based on the current logic we uphold in ReportPortal, the decision should rest with the Test Runner.

Therefore, I wouldn't label this as a bug per se. 
It's more of an `enhancement` request when considering various strategies around handling retry test cases.

Whether or not to incorporate this enhancement is a discussion for the community, as it's crucial to understand if they encounter the same issue. 

Thanks.

**Comment by DzmitryHumianiuk** [comment]
@abhibajaj1987 i've made a quick research, and this gives me the next findings:

> **Take the Last Result** -  (ReportPortal currently uses)The simplest approach is to consider only the last result of the test execution. If a test case is retried three times and fails twice but passes on the third attempt, then the final status is considered "Passed."
> 
> **Majority Rule** - If a test is retried multiple times, the most frequently occurring status could be taken as the final status. For instance, if a test case has been retried five times with statuses [Fail, Pass, Fail, Fail, Pass], the final status would be "Fail" because it appears more frequently.
> 
> **All Must Pass** - In some scenarios, especially where high reliability is required, you may consider a test as "Passed" only if all retries pass. Even a single failure would mark the test as "Failed."
> 
> **First Failure Captured** - Some teams prefer to capture the first failure and report that as the final status, especially if diagnosing transient issues is important. In this strategy, as soon as a failure is encountered, no more retries are conducted, and the test is marked as "Failed."
> 
> **Weighted Approach**- More complex approaches can assign different weights to different attempts and calculate a weighted average or a composite status based on that. This is less common but may be suitable for specific needs.
> 
> **Custom Logic** - Sometimes, custom logic based on the application's unique needs or business rules could be applied.

Based on this, your case seems to be a custom logic, which is opposite to `All Must Pass` in a form of `If at least one Pass`.

This makes me think, that it probably sit on the side of test case stability.
Like comparing strategy `execution this flaky test until we got Passed status` and `execute N times, and consider it Passed in case if it passed at least once`

**Comment by abhibajaj1987** [comment]
@DzmitryHumianiuk , i agree with you on the part where you mentioned that if the test case failed after retrying or passed within a given retry count or if test case eventually failed after retrying it N times, test case should be marked as marked. 

But if you look at the screenshot in the description, test case first passed and then it got failed which is not possible since as per testng retry listener logic, test case would only be retried if test case failed in initial run. 

Order of run seems to be incorrect in report portal. 

Let me brief how the execution happens in testng., say retry max count is 3. 
Retry 1: Failed
Retry 2: failed:
Retry 3: Passed:

Final status should be marked as failed 

Sceanrio 2:
Retry 1: Failed
Retry 2: Passed:

Final status should be marked as failed skipping the retry 3 run since this is desired state.

Also i have some instances in report portal where it is working fine where if test is passed, final status is marked as passed unlike the screenshot where test case failed and then test run was done on it again which is contrary to testng implementation. Looks like some race condition is happening though test is running sequentially. 

Find the screenshot attached to see the difference in RP behaviour.
<img width="1469" alt="image" src="https://github.com/reportportal/reportportal/assets/148849999/bf5b19a9-bbd9-4a7f-b101-2d86b76676b6">

<img width="1469" alt="image" src="https://github.com/reportportal/reportportal/assets/148849999/7d5045f3-629d-4155-ac96-50f741f1b5ef">

**Comment by DzmitryHumianiuk** [comment]
@abhibajaj1987, good catch! I understand what you're pointing out.

This could indeed be due to a race condition when recording and updating the test item as the API receives the "Retry" object. The current implementation seems to have issues with deadlocks in the database when saving the retries, and this inconsistency might stem from workarounds we've implemented to avoid those deadlocks.

Regardless, we'll look into this further.

**Comment by DzmitryHumianiuk** [comment]
this might be co-related https://github.com/reportportal/reportportal/issues/1588

**Comment by abhibajaj1987** [comment]
Also @DzmitryHumianiuk , i made a fix in my report portal client where i am setting retry flag =true only when it is retry test. Earlier ( above screenshots) i was adding retry flag to all test runs which is logically incorrect. 

So, below is the screenshot when retry flag=true only when test is retry. 
<img width="1469" alt="image" src="https://github.com/reportportal/reportportal/assets/148849999/939f5fde-401b-4e2b-8e75-8abf4338f55d">

Now, the question is these skipped cases are counted in total run which is somewhat correct behaviour? 
Can we not club all skipped test in parent pass or fail test so that i see actual count of test being run ( excluding retries) and retry test acts as child to actual parent test case present in either failed or pass status. 

Refer entries in skipped test case. Here, my test cases finally ends up in skip or fail after retry, so there are duplicate entries in skip and fail/pass category
<img width="1474" alt="image" src="https://github.com/reportportal/reportportal/assets/148849999/31355777-7400-4723-8050-b0ab57d15f3c">

If you see in very above screenshot, test first retried 2 times and then it eventually failed which is correct behaviour but however i do not want to show as separate test run and i guess RP handles it but not happening here somehow. Any suggestions?

**Comment by abhibajaj1987** [comment]
@DzmitryHumianiuk is there any update on these 2 issues? 
1. Ignoring the reporting of retried test in skipped test count? 
2. Fixing the issue where race condition is happening due to which PASSED/FAILED test item is reported before skipped during retring?

**Comment by DzmitryHumianiuk** [comment]
@abhibajaj1987 Hi,

For point #1, I'm not sure I fully understand your concern. However, if a separate test item with a 'skipped' status is received by the API, it remains as a separate test item, regardless of its status.

This might also relate to point #2. By design, you should see only 1 test item on the screen with three retries, where 'FAILED' is the final status. What you're currently seeing doesn't seem to be the expected behavior. And I'm not sure where it comes from.

Are you using your custom integration or something officially released by the ReportPortal team?

**Comment by DzmitryHumianiuk** [comment]
@abhibajaj1987 interesting behaviour.
probably we need to take a look into the internal flow how testNG organizes this.
which version on TestNG do you use?
